src/librustc_driver/driver.rs

```diff
@@ -150,16 +150,23 @@ pub fn compile_input(sess: Session,
 
         (outputs, trans, tcx.sess)
     };
-    phase_5_run_llvm_passes(&sess, &trans, &outputs);
-
-    controller_entry_point!(after_llvm,
+    if !sess.targeting_pnacl() {
+        phase_5_run_llvm_passes(&sess, &mut trans, &outputs);
+        controller_entry_point!(after_llvm,
                             sess,
                             CompileState::state_after_llvm(input,
                                                            &sess,
                                                            outdir,
                                                            &trans));
 
-    phase_6_link_output(&sess, &trans, &outputs);
+        phase_6_link_output(&sess, &trans, &outputs);
+    } else {
+        let cid = trans.link.crate_name.clone();
+        link::link_outputs_for_pnacl(&sess,
+                                     &mut trans,
+                                     &outputs,
+                                     &cid[..]);
+    }
```

Comment:
Done. I didn't fold any phases together as you suggested, instead I opted to shuffle the branching to the `phase_*` functions.